// igniteui-angular/esm2020/lib/directives/form-control/form-control.directive.mjs
import { Directive, forwardRef, HostListener, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as i0 from "@angular/core";
export class IgcFormControlDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        /** @hidden @internal */
        this.onChange = () => { };
        /** @hidden @internal */
        this.onTouched = () => { };
    }
    /** @hidden @internal */
    onBlur() {
        this.onTouched();
    }
    /** @hidden @internal */
    listenForValueChange(value) {
        this.onChange(value);
    }
    /** @hidden @internal */
    writeValue(value) {
        if (value) {
            this.elementRef.nativeElement.value = value;
        }
    }
    /** @hidden @internal */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /** @hidden @internal */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /** @hidden @internal */
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
}
IgcFormControlDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgcFormControlDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
IgcFormControlDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgcFormControlDirective, selector: "igc-rating[ngModel],igc-rating[formControlName]", host: { listeners: { "blur": "onBlur()", "igcChange": "listenForValueChange($event.detail)" } }, providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => IgcFormControlDirective),
            multi: true
        }
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgcFormControlDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'igc-rating[ngModel],igc-rating[formControlName]',
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => IgcFormControlDirective),
                            multi: true
                        }
                    ]
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { onBlur: [{
                type: HostListener,
                args: ['blur']
            }], listenForValueChange: [{
                type: HostListener,
                args: ['igcChange', ['$event.detail']]
            }] } });
export class IgcFormsModule {
}
IgcFormsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgcFormsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IgcFormsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.0", ngImport: i0, type: IgcFormsModule, declarations: [IgcFormControlDirective], exports: [IgcFormControlDirective] });
IgcFormsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgcFormsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgcFormsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgcFormControlDirective],
                    exports: [IgcFormControlDirective]
                }]
        }] });
