// igniteui-angular/esm2020/lib/grids/toolbar/common.mjs
import { Component, Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
// eslint-disable-next-line @angular-eslint/directive-selector
export class IgxExcelTextDirective {
}
IgxExcelTextDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxExcelTextDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
IgxExcelTextDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxExcelTextDirective, selector: "[excelText],excel-text", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxExcelTextDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[excelText],excel-text' }]
        }] });
// eslint-disable-next-line @angular-eslint/directive-selector
export class IgxCSVTextDirective {
}
IgxCSVTextDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCSVTextDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
IgxCSVTextDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCSVTextDirective, selector: "[csvText],csv-text", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCSVTextDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[csvText],csv-text' }]
        }] });
/**
 * Provides a way to template the title portion of the toolbar in the grid.
 *
 * @igxModule IgxGridToolbarModule
 * @igxParent IgxGridToolbarComponent
 *
 * @example
 * ```html
 * <igx-grid-toolbar-title>My custom title</igx-grid-toolbar-title>
 * ```
 */
export class IgxGridToolbarTitleComponent {
    constructor() {
        /**
         * Host `class.igx-grid-toolbar__title` binding.
         *
         * @hidden
         * @internal
         */
        this.cssClass = 'igx-grid-toolbar__title';
    }
}
IgxGridToolbarTitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarTitleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
IgxGridToolbarTitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.0", type: IgxGridToolbarTitleComponent, selector: "igx-grid-toolbar-title", host: { properties: { "class.igx-grid-toolbar__title": "this.cssClass" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarTitleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'igx-grid-toolbar-title', template: '<ng-content></ng-content>' }]
        }], propDecorators: { cssClass: [{
                type: HostBinding,
                args: ['class.igx-grid-toolbar__title']
            }] } });
/**
 * Provides a way to template the action portion of the toolbar in the grid.
 *
 * @igxModule IgxGridToolbarModule
 * @igxParent IgxGridToolbarComponent
 *
 * @example
 * ```html
 * <igx-grid-toolbar-actions>
 *  <some-toolbar-action-here />
 * </igx-grid-toolbar-actions>
 * ```
 */
export class IgxGridToolbarActionsComponent {
    constructor() {
        /**
         * Host `class.igx-grid-toolbar__actions` binding.
         *
         * @hidden
         * @internal
         */
        this.cssClass = 'igx-grid-toolbar__actions';
    }
}
IgxGridToolbarActionsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarActionsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
IgxGridToolbarActionsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.0", type: IgxGridToolbarActionsComponent, selector: "igx-grid-toolbar-actions", host: { properties: { "class.igx-grid-toolbar__actions": "this.cssClass" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarActionsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'igx-grid-toolbar-actions', template: '<ng-content></ng-content>' }]
        }], propDecorators: { cssClass: [{
                type: HostBinding,
                args: ['class.igx-grid-toolbar__actions']
            }] } });
export class IgxGridToolbarDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_dir, ctx) {
        return true;
    }
    ;
}
IgxGridToolbarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxGridToolbarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxGridToolbarDirective, selector: "[igxGridToolbar]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridToolbarDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[igxGridToolbar]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
