// igniteui-angular/esm2020/lib/grids/pivot-grid/pivot-grid.module.mjs
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";
import { IgxAccordionModule } from '../../accordion/accordion.module';
import { IgxDragDropModule } from "../../directives/drag-drop/drag-drop.directive";
import { IgxExpansionPanelModule } from "../../expansion-panel/expansion-panel.module";
import { IgxGridModule } from "../grid/grid.module";
import { IgxListModule } from '../../list/list.component';
import { IgxPivotDataSelectorComponent } from "./pivot-data-selector.component";
import { IgxPivotGridComponent } from "./pivot-grid.component";
import { IgxFilterPivotItemsPipe, IgxPivotGridCellStyleClassesPipe, IgxPivotAutoTransform, IgxPivotCellMergingPipe, IgxPivotColumnPipe, IgxPivotGridColumnSortingPipe, IgxPivotGridFilterPipe, IgxPivotGridSortingPipe, IgxPivotRowExpansionPipe, IgxPivotRowPipe } from "./pivot-grid.pipes";
import { IgxPivotHeaderRowComponent } from "./pivot-header-row.component";
import { IgxPivotRowDimensionContentComponent } from "./pivot-row-dimension-content.component";
import { IgxPivotRowDimensionHeaderGroupComponent } from "./pivot-row-dimension-header-group.component";
import { IgxPivotRowDimensionHeaderComponent } from "./pivot-row-dimension-header.component";
import { IgxPivotRowComponent } from "./pivot-row.component";
import { IgxPivotValueChipTemplateDirective } from './pivot-grid.directives';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class IgxPivotGridModule {
}
IgxPivotGridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotGridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IgxPivotGridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotGridModule, declarations: [IgxPivotGridComponent,
        IgxPivotRowComponent,
        IgxPivotHeaderRowComponent,
        IgxPivotRowDimensionContentComponent,
        IgxPivotRowDimensionHeaderComponent,
        IgxPivotRowDimensionHeaderGroupComponent,
        IgxPivotRowPipe,
        IgxPivotRowExpansionPipe,
        IgxPivotAutoTransform,
        IgxPivotColumnPipe,
        IgxPivotGridFilterPipe,
        IgxPivotGridSortingPipe,
        IgxPivotGridColumnSortingPipe,
        IgxPivotCellMergingPipe,
        IgxFilterPivotItemsPipe,
        IgxPivotGridCellStyleClassesPipe,
        IgxPivotDataSelectorComponent,
        IgxPivotValueChipTemplateDirective], imports: [IgxGridModule, IgxExpansionPanelModule, IgxDragDropModule, IgxListModule, IgxAccordionModule], exports: [IgxGridModule,
        IgxPivotGridComponent,
        IgxPivotRowComponent,
        IgxPivotHeaderRowComponent,
        IgxPivotRowDimensionContentComponent,
        IgxPivotRowDimensionHeaderComponent,
        IgxPivotRowDimensionHeaderGroupComponent,
        IgxPivotRowExpansionPipe,
        IgxPivotAutoTransform,
        IgxPivotRowPipe,
        IgxPivotColumnPipe,
        IgxPivotGridFilterPipe,
        IgxPivotGridSortingPipe,
        IgxPivotGridColumnSortingPipe,
        IgxPivotCellMergingPipe,
        IgxFilterPivotItemsPipe,
        IgxPivotGridCellStyleClassesPipe,
        IgxPivotDataSelectorComponent,
        IgxPivotValueChipTemplateDirective] });
IgxPivotGridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotGridModule, imports: [IgxGridModule, IgxExpansionPanelModule, IgxDragDropModule, IgxListModule, IgxAccordionModule, IgxGridModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotGridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        IgxPivotGridComponent,
                        IgxPivotRowComponent,
                        IgxPivotHeaderRowComponent,
                        IgxPivotRowDimensionContentComponent,
                        IgxPivotRowDimensionHeaderComponent,
                        IgxPivotRowDimensionHeaderGroupComponent,
                        IgxPivotRowPipe,
                        IgxPivotRowExpansionPipe,
                        IgxPivotAutoTransform,
                        IgxPivotColumnPipe,
                        IgxPivotGridFilterPipe,
                        IgxPivotGridSortingPipe,
                        IgxPivotGridColumnSortingPipe,
                        IgxPivotCellMergingPipe,
                        IgxFilterPivotItemsPipe,
                        IgxPivotGridCellStyleClassesPipe,
                        IgxPivotDataSelectorComponent,
                        IgxPivotValueChipTemplateDirective,
                    ],
                    exports: [
                        IgxGridModule,
                        IgxPivotGridComponent,
                        IgxPivotRowComponent,
                        IgxPivotHeaderRowComponent,
                        IgxPivotRowDimensionContentComponent,
                        IgxPivotRowDimensionHeaderComponent,
                        IgxPivotRowDimensionHeaderGroupComponent,
                        IgxPivotRowExpansionPipe,
                        IgxPivotAutoTransform,
                        IgxPivotRowPipe,
                        IgxPivotColumnPipe,
                        IgxPivotGridFilterPipe,
                        IgxPivotGridSortingPipe,
                        IgxPivotGridColumnSortingPipe,
                        IgxPivotCellMergingPipe,
                        IgxFilterPivotItemsPipe,
                        IgxPivotGridCellStyleClassesPipe,
                        IgxPivotDataSelectorComponent,
                        IgxPivotValueChipTemplateDirective,
                    ],
                    imports: [IgxGridModule, IgxExpansionPanelModule, IgxDragDropModule, IgxListModule, IgxAccordionModule],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                }]
        }] });
