// igniteui-angular/esm2020/lib/grids/pivot-grid/pivot-grid.directives.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class IgxPivotValueChipTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxPivotValueChipTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotValueChipTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxPivotValueChipTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxPivotValueChipTemplateDirective, selector: "[igxPivotValueChip]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxPivotValueChipTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxPivotValueChip]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
