// igniteui-angular/esm2020/lib/grids/columns/templates.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class IgxFilterCellTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxFilterCellTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxFilterCellTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxFilterCellTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxFilterCellTemplateDirective, selector: "[igxFilterCellTemplate]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxFilterCellTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxFilterCellTemplate]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxCellTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxCellTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCellTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCellTemplateDirective, selector: "[igxCell]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxCell]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxCellValidationErrorDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxCellValidationErrorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellValidationErrorDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCellValidationErrorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCellValidationErrorDirective, selector: "[igxCellValidationError]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellValidationErrorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxCellValidationError]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxCellHeaderTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxCellHeaderTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellHeaderTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCellHeaderTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCellHeaderTemplateDirective, selector: "[igxHeader]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellHeaderTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxHeader]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
/**
 * @hidden
 */
export class IgxCellFooterTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
IgxCellFooterTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellFooterTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCellFooterTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCellFooterTemplateDirective, selector: "[igxFooter]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellFooterTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxFooter]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxCellEditorTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxCellEditorTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellEditorTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCellEditorTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCellEditorTemplateDirective, selector: "[igxCellEditor]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCellEditorTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxCellEditor]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxCollapsibleIndicatorTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxCollapsibleIndicatorTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCollapsibleIndicatorTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxCollapsibleIndicatorTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxCollapsibleIndicatorTemplateDirective, selector: "[igxCollapsibleIndicator]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxCollapsibleIndicatorTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxCollapsibleIndicator]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class IgxSummaryTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    ;
}
IgxSummaryTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxSummaryTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
IgxSummaryTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxSummaryTemplateDirective, selector: "[igxSummary]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxSummaryTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxSummary]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
