// igniteui-angular/esm2020/lib/grids/columns/validators.directive.mjs
import { Directive } from '@angular/core';
import { RequiredValidator, NG_VALIDATORS, MinValidator, MaxValidator, EmailValidator, MinLengthValidator, MaxLengthValidator, PatternValidator } from '@angular/forms';
import * as i0 from "@angular/core";
export class IgxColumnRequiredValidatorDirective extends RequiredValidator {
}
IgxColumnRequiredValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnRequiredValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumnRequiredValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumnRequiredValidatorDirective, selector: "igx-column[required]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumnRequiredValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnRequiredValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[required]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnRequiredValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumnMinValidatorDirective extends MinValidator {
}
IgxColumnMinValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMinValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumnMinValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumnMinValidatorDirective, selector: "igx-column[min]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumnMinValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMinValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[min]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMinValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumnMaxValidatorDirective extends MaxValidator {
}
IgxColumnMaxValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMaxValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumnMaxValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumnMaxValidatorDirective, selector: "igx-column[max]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumnMaxValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMaxValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[max]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMaxValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumnEmailValidatorDirective extends EmailValidator {
}
IgxColumnEmailValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnEmailValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumnEmailValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumnEmailValidatorDirective, selector: "igx-column[email]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumnEmailValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnEmailValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[email]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnEmailValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumnMinLengthValidatorDirective extends MinLengthValidator {
}
IgxColumnMinLengthValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMinLengthValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumnMinLengthValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumnMinLengthValidatorDirective, selector: "igx-column[minlength]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumnMinLengthValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumnMinLengthValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[minlength]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMinLengthValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumMaxLengthValidatorDirective extends MaxLengthValidator {
}
IgxColumMaxLengthValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumMaxLengthValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumMaxLengthValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumMaxLengthValidatorDirective, selector: "igx-column[maxlength]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumMaxLengthValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumMaxLengthValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[maxlength]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumMaxLengthValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
export class IgxColumPatternValidatorDirective extends PatternValidator {
}
IgxColumPatternValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumPatternValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
IgxColumPatternValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.0", type: IgxColumPatternValidatorDirective, selector: "igx-column[pattern]", providers: [{
            provide: NG_VALIDATORS,
            useExisting: IgxColumPatternValidatorDirective,
            multi: true
        }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxColumPatternValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[pattern]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumPatternValidatorDirective,
                            multi: true
                        }]
                }]
        }] });
