// igniteui-angular/esm2020/lib/grids/common/grid-pipes.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxGridFilterConditionPipe, IgxGridTransactionPipe, IgxGridNotGroupedPipe, IgxGridTopLevelColumns, IgxGridCellStylesPipe, IgxGridCellStyleClassesPipe, IgxGridPaginatorOptionsPipe, IgxHasVisibleColumnsPipe, IgxGridRowPinningPipe, IgxGridDataMapperPipe, IgxStringReplacePipe, IgxGridTransactionStatePipe, IgxColumnFormatterPipe, IgxSummaryFormatterPipe, IgxGridAddRowPipe, IgxGridRowClassesPipe, IgxGridRowStylesPipe, IgxGridCellImageAltPipe, } from './pipes';
import * as i0 from "@angular/core";
export * from './pipes';
export class IgxGridPipesModule {
}
IgxGridPipesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridPipesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IgxGridPipesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.0", ngImport: i0, type: IgxGridPipesModule, declarations: [IgxGridFilterConditionPipe,
        IgxGridTransactionPipe,
        IgxGridNotGroupedPipe,
        IgxGridTopLevelColumns,
        IgxGridCellStylesPipe,
        IgxGridCellStyleClassesPipe,
        IgxGridCellImageAltPipe,
        IgxGridPaginatorOptionsPipe,
        IgxHasVisibleColumnsPipe,
        IgxGridRowPinningPipe,
        IgxGridDataMapperPipe,
        IgxStringReplacePipe,
        IgxGridTransactionStatePipe,
        IgxGridAddRowPipe,
        IgxColumnFormatterPipe,
        IgxSummaryFormatterPipe,
        IgxGridRowClassesPipe,
        IgxGridRowStylesPipe], imports: [CommonModule], exports: [IgxGridFilterConditionPipe,
        IgxGridTransactionPipe,
        IgxGridNotGroupedPipe,
        IgxGridTopLevelColumns,
        IgxGridCellStylesPipe,
        IgxGridCellStyleClassesPipe,
        IgxGridCellImageAltPipe,
        IgxGridPaginatorOptionsPipe,
        IgxHasVisibleColumnsPipe,
        IgxGridRowPinningPipe,
        IgxGridDataMapperPipe,
        IgxStringReplacePipe,
        IgxGridTransactionStatePipe,
        IgxGridAddRowPipe,
        IgxColumnFormatterPipe,
        IgxSummaryFormatterPipe,
        IgxGridRowClassesPipe,
        IgxGridRowStylesPipe] });
IgxGridPipesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridPipesModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: IgxGridPipesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        IgxGridFilterConditionPipe,
                        IgxGridTransactionPipe,
                        IgxGridNotGroupedPipe,
                        IgxGridTopLevelColumns,
                        IgxGridCellStylesPipe,
                        IgxGridCellStyleClassesPipe,
                        IgxGridCellImageAltPipe,
                        IgxGridPaginatorOptionsPipe,
                        IgxHasVisibleColumnsPipe,
                        IgxGridRowPinningPipe,
                        IgxGridDataMapperPipe,
                        IgxStringReplacePipe,
                        IgxGridTransactionStatePipe,
                        IgxGridAddRowPipe,
                        IgxColumnFormatterPipe,
                        IgxSummaryFormatterPipe,
                        IgxGridRowClassesPipe,
                        IgxGridRowStylesPipe
                    ],
                    exports: [
                        IgxGridFilterConditionPipe,
                        IgxGridTransactionPipe,
                        IgxGridNotGroupedPipe,
                        IgxGridTopLevelColumns,
                        IgxGridCellStylesPipe,
                        IgxGridCellStyleClassesPipe,
                        IgxGridCellImageAltPipe,
                        IgxGridPaginatorOptionsPipe,
                        IgxHasVisibleColumnsPipe,
                        IgxGridRowPinningPipe,
                        IgxGridDataMapperPipe,
                        IgxStringReplacePipe,
                        IgxGridTransactionStatePipe,
                        IgxGridAddRowPipe,
                        IgxColumnFormatterPipe,
                        IgxSummaryFormatterPipe,
                        IgxGridRowClassesPipe,
                        IgxGridRowStylesPipe
                    ],
                    imports: [
                        CommonModule
                    ]
                }]
        }] });
